### PR TITLE
Define a util trait to handle the integer arithmetic

### PR DESCRIPTION
**Pull request by mrwwheat7**

### CODE REVIEW

File: src/fuse/utils.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
+/// A type cast trait used to do the integer arithmetic.
+pub trait Uint {
+    /// Overflow add.
+    fn overflow_add(self, other: Self) -> Self
+    where
+        Self: std::fmt::Display + num_traits::ops::checked::CheckedAdd,
+    {
+        let result = Self::checked_add(&self, &other);
```
Comment (pwang7) [inline src/fuse/utils.rs]: why not `self.checked_add(other)`?

File: src/fuse/utils.rs [hunk 2]
```diff
@@ -0,0 +1,76 @@
+/// A type cast trait used to do the integer arithmetic.
+pub trait Uint {
+    /// Overflow add.
+    fn overflow_add(self, other: Self) -> Self
```
Comment (pwang7) [inline src/fuse/utils.rs]: `other: &Self`

**Review by pwang7 — CHANGES_REQUESTED** [review]

File: Cargo.toml [hunk 3]
```diff
@@ -7,6 +7,7 @@ edition = "2018"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
+num-traits = "0.2.12"
```
Comment (pwang7) [inline Cargo.toml]: remove it

File: src/fuse/utils.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
+macro_rules! impl_from {
```
Comment (pwang7) [inline src/fuse/utils.rs]: change the macro name to a more meaningful one

File: src/fuse/utils.rs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+impl_from!(i8);
+impl_from!(i16);
+impl_from!(i32);
+impl_from!(i64);
+impl_from!(i128);
+impl_from!(usize);
+impl_from!(isize);
+
+/// A type cast trait used to do the integer arithmetic.
+pub trait Uint<T> {
```
Comment (pwang7) [inline src/fuse/utils.rs]: change the trait name to a more meaningful one

**Review by pwang7 — CHANGES_REQUESTED** [review]

File: Cargo.toml (hunk 3 above)
Comment (mrwwheat7) [reply Cargo.toml]: Done.

File: src/fuse/utils.rs (hunk 4 above)
Comment (mrwwheat7) [reply src/fuse/utils.rs]: Done.

File: src/fuse/utils.rs (hunk 5 above)
Comment (mrwwheat7) [reply src/fuse/utils.rs]: Done.

File: src/fuse/utils.rs [hunk 6]
```diff
@@ -0,0 +1,76 @@
+macro_rules! impl_overflow_arithmetic {
+    ($target: ty) => {
+        impl OverflowArithmetic<$target> for $target {
+            #[inline]
+            fn overflow_add(self, other: $target) -> Self {
+                let (res, overflow) = self.overflowing_add(other);
+                debug_assert!(
+                    !overflow,
+                    "number = {} add number = {} overflowing",
```
Comment (pzheng1025) [inline src/fuse/utils.rs]: Better to print parameter type here so that it is easier to know which type is overflowed.

Comment (pzheng1025) [reply src/fuse/utils.rs]: like number = 5(u32) add number = XXX overflow ...

Comment (mrwwheat7) [reply src/fuse/utils.rs]: Done.

**Review by pzheng1025 — APPROVED** [review]

**Review by pwang7 — APPROVED** [review]

**Pull request merged**